File: benchmark/comparisons/script/scripts/binary_tree.lua
-- binary_tree.lua -- build a balanced binary tree of depth arg[1], then
-- recursively count its nodes. Same shape as binary_tree.pg / .py.

-- Same shape as the .pg / .py versions: each node stores `d` (depth) so
-- leaves can be detected via int==int. See binary_tree.py for the reasoning.

local function makeNode(depth)
    local n = { d = depth }
    if depth > 0 then
        n.left = makeNode(depth - 1)
        n.right = makeNode(depth - 1)
    end
    return n
end

local function countNodes(node)
    if node.d == 0 then return 1 end
    return 1 + countNodes(node.left) + countNodes(node.right)
end

local count = tonumber(arg[1])

local t0 = os.clock()
local root = makeNode(count)
local total = countNodes(root)
local elapsed_ns = (os.clock() - t0) * 1e9

print(math.floor(elapsed_ns))

File: benchmark/comparisons/script/scripts/binary_tree.py
# binary_tree.py -- build a balanced binary tree of depth argv[1], then
# recursively count its nodes. Same shape as binary_tree.pg / .lua.

import sys
import time


class Node:
    # `d` (depth) is stored so countNodes can distinguish leaves via int==int.
    # PgScript can't compare instance==int or read unset fields, so all three
    # languages carry the depth field for a fair comparison.
    __slots__ = ("d", "left", "right")

    def __init__(self, depth):
        self.d = depth
        if depth > 0:
            self.left = Node(depth - 1)
            self.right = Node(depth - 1)


def count_nodes(node):
    if node.d == 0:
        return 1
    return 1 + count_nodes(node.left) + count_nodes(node.right)


count = int(sys.argv[1])

t0 = time.perf_counter_ns()
root = Node(count)
total = count_nodes(root)
elapsed = time.perf_counter_ns() - t0

print(elapsed)
